devito/passes/iet/definitions.py

```diff
@@ -448,15 +450,34 @@ def _map_function_on_high_bw_mem(self, site, obj, storage, devicerm, read_only=F
         `_map_array_on_high_bw_mem` is that the former triggers a data transfer to
         synchronize the host and device copies, while the latter does not.
         """
-        mmap = self.lang._map_to(obj)
-
         if read_only is False:
+            if is_gpu_create(obj, self.gpu_create):
+                mmap = self.lang._map_alloc(obj)
+
+                cdims = make_zero_init(obj)
+                eq = Eq(obj[cdims], 0)
+
+                irs, _ = self.rcompile(eq)
+
+                init = irs.iet.body.body[0]
+
+                name = self.sregistry.make_name(prefix='init')
+                efunc = make_callable(name, init)
+                init = Call(name, efunc.parameters)
+
+                mmap = (mmap, init)
+            else:
+                mmap = self.lang._map_to(obj)
+                efunc = ()
+
```

Comment:
Conceptual question. If the Function is initialized on device, the "assumption" is that that function can be considered "temporary" so should the `_map_update` be optional as well, i.e never sent back to host just freed

Reply:
I think there might be cases when it's temporary and others in which it's not. The assumption is that these Functions are zero-initialised and therefore do not need to be copied from host, but their final value might be relevant to the user. The example I can think of is the computed wavefield in a forward solve. It would be good to have separate levers to control this, though

Reply:
> should the _map_update be optional as well

I think it's impossible to say in general. If I'm writing a test, I for sure want an update. Maybe an application code won't care. I agree in the future we may need further opt-options to e.g. distinguish between what is to be updated back and what not